provider/equinix/environ.go

```diff
@@ -78,31 +81,100 @@ func (e *environ) Config() *config.Config {
+	validator.RegisterVocabulary(constraints.Arch, []string{arch.AMD64, arch.ARM64, arch.I386, arch.PPC64EL})
+	return validator, nil
 }
 
 func (e *environ) ControllerInstances(ctx context.ProviderCallContext, controllerUUID string) ([]instance.Id, error) {
-	return nil, errors.NewNotImplemented(nil, "not implemented")
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-is-controller": "true", "juju-controller-uuid": controllerUUID})
+	if err != nil {
+		return nil, err
+	}
+	instanceIDs := make([]instance.Id, len(insts))
+	for i, inst := range insts {
+		instanceIDs[i] = inst.Id()
+	}
+	return instanceIDs, nil
 }
 
 func (e *environ) Create(ctx context.ProviderCallContext, args environs.CreateParams) error {
 	return nil
 }
 
 func (e *environ) Destroy(ctx context.ProviderCallContext) error {
-	return errors.NewNotImplemented(nil, "not implemented")
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, inst := range insts {
+		if _, err = e.equinixClient.Devices.Delete(string(inst.Id()), true); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return common.Destroy(e, ctx)
 }
 
 func (e *environ) DestroyController(ctx context.ProviderCallContext, controllerUUID string) error {
-	return errors.NewNotImplemented(nil, "not implemented")
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-controller-uuid": controllerUUID})
+	if err != nil {
+		return err
+	}
+
+	for _, inst := range insts {
+		if _, err = e.equinixClient.Devices.Delete(string(inst.Id()), true); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return e.Destroy(ctx)
 }
 
 func (e *environ) InstanceTypes(context.ProviderCallContext, constraints.Value) (instances.InstanceTypesWithCostMetadata, error) {
-	panic(errors.NewNotImplemented(nil, "not implemented"))
+	i := instances.InstanceTypesWithCostMetadata{}
+	instances, err := e.supportedInstanceTypes()
+	if err != nil {
+		return i, errors.Trace(err)
+	}
+
+	i.InstanceTypes = instances
+	return i, nil
 }
 
 func (e *environ) Instances(ctx context.ProviderCallContext, ids []instance.Id) ([]instances.Instance, error) {
-	panic(errors.NewNotImplemented(nil, "not implemented"))
+	toReturn := make([]instances.Instance, len(ids))
+	var missingInstanceCount int
+
+	tags := set.NewStrings("juju-model-uuid=" + e.Config().UUID())
+
+	for i, id := range ids {
+		d, resp, err := e.equinixClient.Devices.Get(string(id), nil)
+		if err != nil && resp != nil && resp.Request.Response.StatusCode == http.StatusNotFound {
+			logger.Warningf("instance %s not found", string(id))
+			missingInstanceCount = missingInstanceCount + 1
+			continue
+		} else if err != nil {
+			return nil, errors.Annotatef(err, "looking up device with ID %q", id)
+		}
+
+		deviceTags := set.NewStrings(d.Tags...)
+		if !tags.Intersection(deviceTags).IsEmpty() {
```

Comment:
You should increment `missingInstanceCount` if you don't match the tags. I suggest inverting the check as follows:

```go
deviceTags := set.NewStrings(d.Tags...)
if tags.Intersection(deviceTags).IsEmpty() {
   missingInstanceCount++
   continue
}
toReturn[i] = &equinixDevice{e, d}
```